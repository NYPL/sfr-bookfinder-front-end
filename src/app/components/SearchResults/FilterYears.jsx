import React from 'react';
import PropTypes from 'prop-types';
import { initialSearchQuery, searchQueryPropTypes } from '../../stores/InitialState';
import TextInput from '../Form/TextInput';

import { yearsType, errorMessagesText } from '../../constants/labels';

const getYearsFilter = (searchQuery) => {
  const yearsValues = {};
  Object.keys(yearsType).forEach((yearType) => {
    const yearValue = searchQuery && searchQuery.filters && searchQuery.filters.find(filter => filter.field === 'years');
    yearsValues[yearType] = yearValue && yearValue.value[yearType] ? Number(yearValue.value[yearType]) : '';
  });
  return yearsValues;
};

class FilterYears extends React.Component {
  constructor(props) {
    super(props);
    this.state = { ...getYearsFilter(props.searchQuery), ...{ errorMessage: {}, error: {} } };
    this.onChangeYear = this.onChangeYear.bind(this);
  }

  componentWillReceiveProps(nextProps) {
    this.setState(prevState => ({ ...getYearsFilter(nextProps.searchQuery), ...prevState }));
  }

  onChangeYear(e, yearType) {
    const val = e.target.value && Number(e.target.value);
    // TODO: errors control UI
    const obj = {};
    obj[yearType] = val;
    this.setState(state => Object.assign({}, state, obj));
    this.props.onChange({ ...{ start: this.state.start, end: this.state.end }, ...obj });
  }

  onBlur(e, yearType) {
    const val = e.target.value && Number(e.target.value);
    const errorMessage = {};
    const error = {};
    const errorMessageText = errorMessagesText.invalidDate;
    if (yearType === 'start') {
      if (this.state.end && val && val > Number(this.state.end)) {
        errorMessage[yearType] = errorMessageText;
        error[yearType] = true;
      } else {
        errorMessage[yearType] = '';
        error[yearType] = false;
      }
    } else if (yearType === 'end') {
      if (this.state.start && val && val < Number(this.state.start)) {
        errorMessage[yearType] = errorMessageText;
        error[yearType] = true;
      } else {
        errorMessage[yearType] = '';
        error[yearType] = false;
      }
    }
    this.setState({ errorMessage, error });
    if (!error.end && !error.start) {
      this.props.onError({ errorMsg: '', error: false });
    } else {
      this.props.onError({ errorMsg: errorMessageText, error: true });
    }
  }

  render() {
    return (
      <div className="grid-container padding-0">
        <div className={this.props.className}>
          {Object.keys(yearsType).map(yearType => (
            <TextInput
              className={this.props.inputClassName}
              ariaLabel={`Search for ${yearsType[yearType]} date`}
              labelClass=""
              id={`filters.years.${yearType}`}
              key={`filters.years.${yearType}`}
              type="number"
              inputClass={this.state.error[yearType] ? 'usa-input usa-input--error' : 'usa-input'}
              name={`filters.years.${yearType}`}
              onChange={e => this.onChangeYear(e, yearType)}
              onBlur={e => this.onBlur(e, yearType)}
              label={yearsType[yearType]}
              value={this.state[yearType]}
              // errorMessage={this.state.errorMessage[yearType]}
              // max={yearType === 'start' ? this.state.end : null}
              // min={yearType === 'end' ? this.state.start : null}
            />
          ))}
        </div>
      </div>
    );
  }
}

FilterYears.propTypes = {
  searchQuery: searchQueryPropTypes,
  onChange: PropTypes.func,
  onError: PropTypes.func,
  inputClassName: PropTypes.string,
  className: PropTypes.string,
};

FilterYears.defaultProps = {
  searchQuery: initialSearchQuery,
  onChange: () => {},
  onError: () => {},
  inputClassName: 'tablet:grid-col padding-right-0 padding-top-2',
  className: 'grid-row grid-gap',
};

export default FilterYears;
