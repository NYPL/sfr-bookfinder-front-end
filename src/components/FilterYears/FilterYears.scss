.year-input-group {
    display: flex;
    div { 
        margin-right: var(--space-s)
    }
}