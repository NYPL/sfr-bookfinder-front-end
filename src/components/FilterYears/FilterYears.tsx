import React from "react";
import * as DS from "@nypl/design-system-react-components";

import { Filter } from "~/src/types/SearchQuery";

/**
 * Year Filters
 * Can be passed as a form or as a fieldset
 *
 *
 * @param props
 */
const FilterYears: React.FC<{
  startFilter: Filter;
  endFilter: Filter;
  onDateChange: (
    e: React.FormEvent<HTMLInputElement>,
    isStart: boolean
  ) => void;
  // The date range error to show.
  // If no error should be shown, this should be an empty string
  dateRangeError?: string;
  onSubmit?: () => void;
}> = (props) => {
  const { startFilter, endFilter, onDateChange, dateRangeError, onSubmit } =
    props;

  const changeDate = (
    isStart: boolean,
    e: React.FormEvent<HTMLInputElement>
  ) => {
    onDateChange(e, isStart);
  };

  // FilterYears can either be a form on its own (widescreen sidebar) or it can be a part of a larger form (advanced search)
  // If it is a part of a larger form, error checking should happen on form submit, and the error should appear at the top of the form rather than
  // next to the FilterYears inputs
  if (dateRangeError && !onSubmit) {
    console.warn(
      "Found a dateRangeError but no onSubmit.  Errors should be shown at the top of the form when this is used as a fieldset."
    );
  }

  return (
    <fieldset className="date-range">
      <legend>Publication Year</legend>
      <div className="year-input-group">
        <div className="year-input">
          <DS.Label id="date-from-label" htmlFor="date-filter-from">
            From
          </DS.Label>
          <DS.Input
            type={DS.InputTypes.number}
            id="date-filter-from"
            attributes={{ "aria-labelledby": "date-from-label" }}
            value={startFilter ? startFilter.value : ""}
            onChange={(e) => changeDate(true, e)}
          />
          <DS.HelperErrorText isError={false}> EX. 1901 </DS.HelperErrorText>
        </div>
        <div className="year-input">
          <DS.Label id="date-to-label" htmlFor="date-filter-to">
            To
          </DS.Label>
          <DS.Input
            attributes={{ "aria-labelledby": "date-to-label" }}
            type={DS.InputTypes.number}
            id="date-filter-to"
            value={endFilter ? endFilter.value : ""}
            onChange={(e) => changeDate(false, e)}
          />
          <DS.HelperErrorText isError={false}> EX. 2000 </DS.HelperErrorText>
        </div>
      </div>
      {dateRangeError && (
        <DS.HelperErrorText isError={true}>{dateRangeError}</DS.HelperErrorText>
      )}
      {onSubmit && (
        <DS.Button
          id="year-filter-button"
          type="button"
          onClick={() => onSubmit()}
        >
          Apply
        </DS.Button>
      )}
    </fieldset>
  );
};

export default FilterYears;
