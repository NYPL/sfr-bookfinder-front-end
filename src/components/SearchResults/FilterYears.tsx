import React from "react";
import {
  initialSearchQuery,
  searchQueryPropTypes,
} from "../../stores/InitialState";
import TextInput from "../Form/TextInput";

import { yearsType, errorMessagesText } from "../../constants/labels";

const getYearsFilter = (searchQuery: any) => {
  const yearsValues = {};
  Object.keys(yearsType).forEach((yearType) => {
    const yearValue =
      searchQuery &&
      searchQuery.filters &&
      searchQuery.filters.find((filter: any) => filter.field === "years");
    // @ts-expect-error ts-migrate(7053) FIXME: Element implicitly has an 'any' type because expre... Remove this comment to see the full error message
    yearsValues[yearType] =
      yearValue && yearValue.value[yearType]
        ? Number(yearValue.value[yearType])
        : "";
  });
  return yearsValues;
};

type OwnFilterYearsProps = {
  searchQuery?: searchQueryPropTypes;
  onChange?: (...args: any[]) => any;
  onError?: (...args: any[]) => any;
  inputClassName?: string;
  className?: string;
};

type FilterYearsState = any;

type FilterYearsProps = OwnFilterYearsProps & typeof FilterYears.defaultProps;

class FilterYears extends React.Component<FilterYearsProps, FilterYearsState> {
  static defaultProps = {
    searchQuery: initialSearchQuery,
    onChange: () => {},
    onError: () => {},
    inputClassName: "tablet:grid-col padding-right-0 padding-top-2",
    className: "grid-row grid-gap",
  };

  constructor(props: FilterYearsProps) {
    super(props);
    this.state = {
      ...getYearsFilter(props.searchQuery),
      ...{ errorMessage: {}, error: {} },
    };
    this.onChangeYear = this.onChangeYear.bind(this);
  }

  componentWillReceiveProps(nextProps: FilterYearsProps) {
    this.setState((prevState: any) => ({
      ...getYearsFilter(nextProps.searchQuery),
      ...prevState,
    }));
  }

  onChangeYear(e: any, yearType: any) {
    const val = e.target.value && Number(e.target.value);
    // TODO: errors control UI
    const obj = {};
    // @ts-expect-error ts-migrate(7053) FIXME: Element implicitly has an 'any' type because expre... Remove this comment to see the full error message
    obj[yearType] = val;
    this.setState((state: any) => Object.assign({}, state, obj));
    this.props.onChange({
      ...{ start: this.state.start, end: this.state.end },
      ...obj,
    });
  }

  onBlur(e: any, yearType: any) {
    const val = e.target.value && Number(e.target.value);
    const errorMessage = {};
    const error = {};
    const errorMessageText = errorMessagesText.invalidDate;
    if (yearType === "start") {
      if (this.state.end && val && val > Number(this.state.end)) {
        // @ts-expect-error ts-migrate(7053) FIXME: Element implicitly has an 'any' type because expre... Remove this comment to see the full error message
        errorMessage[yearType] = errorMessageText;
        // @ts-expect-error ts-migrate(7053) FIXME: Element implicitly has an 'any' type because expre... Remove this comment to see the full error message
        error[yearType] = true;
      } else {
        // @ts-expect-error ts-migrate(7053) FIXME: Element implicitly has an 'any' type because expre... Remove this comment to see the full error message
        errorMessage[yearType] = "";
        // @ts-expect-error ts-migrate(7053) FIXME: Element implicitly has an 'any' type because expre... Remove this comment to see the full error message
        error[yearType] = false;
      }
    } else if (yearType === "end") {
      if (this.state.start && val && val < Number(this.state.start)) {
        // @ts-expect-error ts-migrate(7053) FIXME: Element implicitly has an 'any' type because expre... Remove this comment to see the full error message
        errorMessage[yearType] = errorMessageText;
        // @ts-expect-error ts-migrate(7053) FIXME: Element implicitly has an 'any' type because expre... Remove this comment to see the full error message
        error[yearType] = true;
      } else {
        // @ts-expect-error ts-migrate(7053) FIXME: Element implicitly has an 'any' type because expre... Remove this comment to see the full error message
        errorMessage[yearType] = "";
        // @ts-expect-error ts-migrate(7053) FIXME: Element implicitly has an 'any' type because expre... Remove this comment to see the full error message
        error[yearType] = false;
      }
    }
    this.setState({ errorMessage, error });
    // @ts-expect-error ts-migrate(2339) FIXME: Property 'end' does not exist on type '{}'.
    if (!error.end && !error.start) {
      this.props.onError({ errorMsg: "", error: false });
    } else {
      this.props.onError({ errorMsg: errorMessageText, error: true });
    }
  }

  render() {
    return (
      <div className="grid-container padding-0">
        <div className={this.props.className}>
          {Object.keys(yearsType).map((yearType) => (
            <TextInput
              className={this.props.inputClassName}
              // @ts-expect-error ts-migrate(7053) FIXME: Element implicitly has an 'any' type because expre... Remove this comment to see the full error message
              ariaLabel={`Search for ${yearsType[yearType]} date`}
              labelClass=""
              id={`filters.years.${yearType}`}
              key={`filters.years.${yearType}`}
              type="number"
              inputClass={
                this.state.error[yearType]
                  ? "usa-input usa-input--error"
                  : "usa-input"
              }
              name={`filters.years.${yearType}`}
              onChange={(e: any) => this.onChangeYear(e, yearType)}
              onBlur={(e: any) => this.onBlur(e, yearType)}
              // @ts-expect-error ts-migrate(7053) FIXME: Element implicitly has an 'any' type because expre... Remove this comment to see the full error message
              label={yearsType[yearType]}
              value={this.state[yearType]}
              // errorMessage={this.state.errorMessage[yearType]}
              // max={yearType === 'start' ? this.state.end : null}
              // min={yearType === 'end' ? this.state.start : null}
            />
          ))}
        </div>
      </div>
    );
  }
}

export default FilterYears;
