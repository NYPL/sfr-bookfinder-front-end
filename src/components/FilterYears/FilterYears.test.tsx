import React from "react";
import FilterYears from "./FilterYears";
import { mount, shallow } from "enzyme";

describe("Filter Years with no apply button", () => {
  const dateChange = jest.fn();

  test("Renders filter years input", () => {
    const wrapper = shallow(
      <FilterYears
        dateFilters={{ start: 2000, end: 3000 }}
        onDateChange={dateChange}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  test("The inputs are prepopulated with passed-in numbers", () => {
    const wrapper = mount(
      <FilterYears
        dateFilters={{ start: 2000, end: 3000 }}
        onDateChange={dateChange}
      />
    );
    expect(wrapper.find("#date-filter-from").hostNodes().props().value).toEqual(
      2000
    );
    expect(wrapper.find("#date-filter-to").hostNodes().props().value).toEqual(
      3000
    );
  });
  test("Calls dateChange function when input changes", () => {
    const wrapper = mount(
      <FilterYears
        dateFilters={{ start: 2000, end: 3000 }}
        onDateChange={dateChange}
      />
    );
    wrapper.find("#date-filter-from").hostNodes().simulate("change", 2001);
    expect(dateChange).toHaveBeenCalledTimes(1);
    wrapper.find("#date-filter-to").hostNodes().simulate("change", 2001);
    expect(dateChange).toHaveBeenCalledTimes(2);
  });
});

describe("Filter Years with apply button", () => {
  const dateChange = jest.fn();
  const submit = jest.fn();

  test("Renders Filter Years with Apply button", () => {
    const wrapper = shallow(
      <FilterYears
        dateFilters={{ start: 30000, end: 2000 }}
        onDateChange={dateChange}
        onSubmit={submit}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  test("Calls submit without validation when apply button is clicked", () => {
    const wrapper = mount(
      <FilterYears
        dateFilters={{ start: 30000, end: 2000 }}
        onDateChange={dateChange}
        onSubmit={submit}
      />
    );
    expect(wrapper.exists("#year-filter-button")).toBeTruthy();

    wrapper.find("#year-filter-button").hostNodes().simulate("click");
    expect(submit).toBeCalledTimes(1);
    expect(wrapper.find({ isError: true }).length).toEqual(0);
  });
});

describe("Filter Years with error", () => {
  const spy = jest.spyOn(console, "warn");
  test("Shows error message with passed string", () => {
    const wrapper = mount(
      <FilterYears
        dateFilters={{ start: 2000, end: 3000 }}
        onDateChange={jest.fn()}
        dateRangeError="error"
        onSubmit={jest.fn}
      />
    );
    expect(wrapper.find({ isError: true }).text()).toEqual("error");
  });

  test("Throws warning when dateRangeError is passed without onSubmit", () => {
    shallow(
      <FilterYears
        dateFilters={{ start: 2000, end: 3000 }}
        onDateChange={jest.fn()}
        dateRangeError="error"
      />
    );
    expect(spy).toHaveBeenCalledTimes(1);
    spy.mockRestore();
  });
});
