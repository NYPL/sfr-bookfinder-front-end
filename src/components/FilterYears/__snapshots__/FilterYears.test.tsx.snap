// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter Years with apply button Renders Filter Years with Apply button 1`] = `
<fieldset>
  <legend
    className="usa-legend font-sans-lg sub-legend"
  >
    Publication Year
  </legend>
  <Label
    htmlFor="date-filter-from"
    id="date-from-label"
  >
    From
  </Label>
  <Input
    id="date-filter-from"
    onChange={[Function]}
    type="number"
    value={30000}
  />
  <HelperErrorText
    isError={false}
  >
     EX. 1901 
  </HelperErrorText>
  <Label
    htmlFor="date-filter-to"
    id="date-to-label"
  >
    To
  </Label>
  <Input
    id="date-filter-to"
    onChange={[Function]}
    type="number"
    value={2000}
  />
  <HelperErrorText
    isError={false}
  >
     EX. 2000 
  </HelperErrorText>
  <Button
    id="year-filter-button"
    mouseDown={false}
    onClick={[Function]}
    type="button"
  >
    Apply
  </Button>
</fieldset>
`;

exports[`Filter Years with no apply button Renders filter years input 1`] = `
<fieldset>
  <legend
    className="usa-legend font-sans-lg sub-legend"
  >
    Publication Year
  </legend>
  <Label
    htmlFor="date-filter-from"
    id="date-from-label"
  >
    From
  </Label>
  <Input
    id="date-filter-from"
    onChange={[Function]}
    type="number"
    value={2000}
  />
  <HelperErrorText
    isError={false}
  >
     EX. 1901 
  </HelperErrorText>
  <Label
    htmlFor="date-filter-to"
    id="date-to-label"
  >
    To
  </Label>
  <Input
    id="date-filter-to"
    onChange={[Function]}
    type="number"
    value={3000}
  />
  <HelperErrorText
    isError={false}
  >
     EX. 2000 
  </HelperErrorText>
</fieldset>
`;
